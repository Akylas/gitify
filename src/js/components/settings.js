import React from 'react';
import { connect } from 'react-redux';
import Toggle from 'react-toggle';

const ipcRenderer = window.require('electron').ipcRenderer;

import { fetchNotifications, updateSetting, logout } from '../actions';

export class SettingsPage extends React.Component {
  componentWillReceiveProps(nextProps) {
    if (nextProps.settings.get('participating') !== this.props.settings.get('participating')) {
      this.props.fetchNotifications();
    }
  }

  logout() {
    this.props.logout();
    this.context.router.replace('/login');
    ipcRenderer.send('update-icon', 'IconPlain');
  }

  checkForUpdates() {
    ipcRenderer.send('check-update');
  }

  appQuit() {
    ipcRenderer.send('app-quit');
  }

  render() {
    const appVersion = require('../../../package.json').version;
    const { settings } = this.props;

    return (
      <div className="container-fluid main-container settings">
        <ul className="nav nav-pills">
          <li className="nav-item">
            <a className="nav-link" onClick={this.checkForUpdates}>
              <i className="fa fa-cloud-download" /> Update
            </a>
          </li>
          <li className="nav-item">
            <a className="nav-link" onClick={this.logout.bind(this)}>
              <i className="fa fa-sign-out" /> Logout
            </a>
          </li>
          <li className="nav-item">
            <a className="nav-link" onClick={this.appQuit}>
              <i className="fa fa-power-off" /> Quit Gitify
            </a>
          </li>
        </ul>

        <div className="row setting">
          <div className="col-xs-8">Show only participating</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={settings.get('participating')}
              onChange={(evt) => this.props.updateSetting('participating', evt.target.checked)} />
          </div>
        </div>
        <div className="row setting">
          <div className="col-xs-8">Play sound</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={settings.get('playSound')}
              onChange={(evt) => this.props.updateSetting('playSound', evt.target.checked)} />
          </div>
        </div>
        <div className="row setting">
          <div className="col-xs-8">Show notifications</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={settings.get('showNotifications')}
              onChange={(evt) => this.props.updateSetting('showNotifications', evt.target.checked)} />
          </div>
        </div>

        <div className="row setting">
          <div className="col-xs-8">On Click, Mark as Read</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={settings.get('markOnClick')}
              onChange={(evt) => this.props.updateSetting('markOnClick', evt.target.checked)} />
          </div>
        </div>

        <div className="row setting">
          <div className="col-xs-8">Open at startup</div>
          <div className="col-xs-4">
            <Toggle
              defaultChecked={settings.get('openAtStartup')}
              onChange={(evt) => this.props.updateSetting('openAtStartup', evt.target.checked)} />
          </div>
        </div>

        <div className="row footer">
          <div className="col-xs-6 text-left">Made with <span className="heart">‚ù§</span> in Brighton.</div>
          <div className="col-xs-6 text-right">Gitify - Version: {appVersion}</div>
        </div>
      </div>
    );
  }
};

SettingsPage.contextTypes = {
  router: React.PropTypes.object.isRequired
};

function mapStateToProps(state) {
  return {
    settings: state.settings
  };
};

export default connect(mapStateToProps, {
  updateSetting,
  fetchNotifications,
  logout
})(SettingsPage);
