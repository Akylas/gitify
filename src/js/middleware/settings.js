const ipcRenderer = window.require('electron').ipcRenderer;
import { UPDATE_SETTING } from '../actions';

export default store => next => action => {

  switch (action.type) {
    case UPDATE_SETTING:
      if (action.setting === 'openAtStartup') {
        // ipcRenderer.send('app-quit');
        if (action.value) {
          ipcRenderer.send('startup-enable');
        } else {
          ipcRenderer.send('startup-disable');
        }
      }
  }

  return next(action);
};
