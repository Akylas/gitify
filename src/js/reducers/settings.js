import { UPDATE_SETTING, HAS_STARRED } from '../actions';

const initialState = {
  participating: false,
  playSound: true,
  showNotifications: true,
  markOnClick: false,
  openAtStartup: false,
  hasStarred: false
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case UPDATE_SETTING:
      return {
        ...state,
        [action.setting]: action.value
      };
    case HAS_STARRED.SUCCESS:
      return {
        ...state,
        hasStarred: true
      };
    case HAS_STARRED.FAILURE:
      return {
        ...state,
        hasStarred: false
      };
    default:
      return state;
  }
};
