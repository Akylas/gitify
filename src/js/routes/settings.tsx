import * as React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

const { remote } = require('electron');

import { AppState, SettingsState } from '../../types/reducers';
import { fetchNotifications, updateSetting, logout } from '../actions';
import { FieldCheckbox } from '../components/ui/checkbox';
import { updateTrayIcon } from '../utils/comms';

interface IFieldRadio {
  name: string;
  label: string;
  value: string;
  checked: boolean;
  onChange: any;
}
const FieldRadio = (props: IFieldRadio) => {
  return (
    <div>
      <input
        type="radio"
        id={`${props.name}-${props.value}`}
        name={props.name}
        value={props.value}
        checked={props.checked}
        onChange={evt => props.onChange(props.name, evt.target.value)}
      />
      <label htmlFor={`${props.name}-${props.value}`}>{props.label}</label>
    </div>
  );
};

const Wrapper = styled.div`
  padding: 2rem;
`;

const Header = styled.div`
  margin: 0 0 1rem;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

const Footer = styled.div`
  margin: 1rem 0;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

const Title = styled.h3`
  margin: 0;
  font-weight: 500;
`;

const ButtonClose = styled.button`
  border: 0;
  padding: 0.25rem;
  font-size: 2rem;
  font-weight: 500;

  &:hover {
    cursor: pointer;
    color: ${props => props.theme.primary};
  }
`;

const ButtonLogout = styled.button`
  border: 0;
  padding: 0.25rem 0.5rem;
  text-transform: uppercase;
  font-size: 0.8rem;
  font-weight: 500;
  border: 1px solid ${props => props.theme.danger};
  border-radius: ${props => props.theme.borderRadius};

  &:hover {
    background-color: ${props => props.theme.danger};
    color: white;
  }
`;

interface IProps {
  hasMultipleAccounts: boolean;
  fetchNotifications: () => any;
  logout: () => any;
  settings: SettingsState;
  updateSetting: any;
  history: any;
}

export class SettingsRoute extends React.Component<IProps> {
  constructor(props) {
    super(props);

    this.state = {
      participating: props.settings.participating,
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (props.settings.participating !== state.participating) {
      props.fetchNotifications();
    }

    return {
      participating: props.settings.participating,
    };
  }

  logout() {
    this.props.logout();
    this.props.history.goBack();
    updateTrayIcon();
  }

  render() {
    const { hasMultipleAccounts, settings } = this.props;

    return (
      <Wrapper>
        <Header>
          <Title>Settings</Title>

          <ButtonClose
            aria-label="Close Settings"
            onClick={() => this.props.history.goBack()}
          >
            &times;
          </ButtonClose>
        </Header>
        <FieldCheckbox
          name="showOnlyParticipating"
          label="Show only participating"
          checked={settings.participating}
          onChange={evt =>
            this.props.updateSetting('participating', evt.target.checked)
          }
        />
        <FieldCheckbox
          name="playSound"
          label="Play sound"
          checked={settings.playSound}
          onChange={evt =>
            this.props.updateSetting('playSound', evt.target.checked)
          }
        />
        <FieldCheckbox
          name="showNotifications"
          label="Show notifications"
          checked={settings.showNotifications}
          onChange={evt =>
            this.props.updateSetting('showNotifications', evt.target.checked)
          }
        />
        <FieldCheckbox
          name="onClickMarkAsRead"
          label="On Click, Mark as Read"
          checked={settings.markOnClick}
          onChange={evt =>
            this.props.updateSetting('markOnClick', evt.target.checked)
          }
        />
        <FieldCheckbox
          name="openAtStartUp"
          label="Open at startup"
          checked={settings.openAtStartup}
          onChange={evt =>
            this.props.updateSetting('openAtStartup', evt.target.checked)
          }
        />
        <div>
          <form>
            <FieldRadio
              name="showAppIcon"
              value="both"
              checked={settings.showAppIcon === 'both'}
              label="Both Icons"
              onChange={this.props.updateSetting}
            />
            <FieldRadio
              name="showAppIcon"
              value="tray"
              checked={settings.showAppIcon === 'tray'}
              label="Tray Icon"
              onChange={this.props.updateSetting}
            />
            <FieldRadio
              name="showAppIcon"
              value="dock"
              checked={settings.showAppIcon === 'dock'}
              label="Dock Icon"
              onChange={this.props.updateSetting}
            />
          </form>
        </div>

        <Footer>
          <small>Version: {remote.app.getVersion()}</small>
          <ButtonLogout aria-label="Logout" onClick={() => this.logout()}>
            {hasMultipleAccounts ? 'Logout from all accounts' : 'Logout'}
          </ButtonLogout>
        </Footer>
      </Wrapper>
    );
  }
}

export function mapStateToProps(state: AppState) {
  const enterpriseAccounts = state.auth.enterpriseAccounts;
  const isGitHubLoggedIn = state.auth.token !== null;
  const hasMultipleAccounts = isGitHubLoggedIn
    ? enterpriseAccounts.length > 0
    : enterpriseAccounts.length > 1;

  return {
    settings: state.settings,
    hasMultipleAccounts,
  };
}

export default connect(mapStateToProps, {
  updateSetting,
  fetchNotifications,
  logout,
})(SettingsRoute);
