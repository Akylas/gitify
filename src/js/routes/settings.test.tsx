import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

const { ipcRenderer } = require('electron');

import { SettingsRoute, mapStateToProps } from './settings';
import {
  SettingsState,
  AppState,
  AuthState,
  EnterpriseAccount,
} from '../../types/reducers';

describe('routes/settings.tsx', () => {
  const props = {
    updateSetting: jest.fn(),
    fetchNotifications: jest.fn(),
    hasMultipleAccounts: false,
    logout: jest.fn(),
    history: {
      goBack: jest.fn(),
    },
    settings: {
      participating: false,
      playSound: true,
      showNotifications: true,
      markOnClick: false,
      openAtStartup: false,
      hasStarred: false,
      showAppIcon: 'both',
    },
  };

  beforeEach(function() {
    spyOn(ipcRenderer, 'send');

    props.updateSetting.mockReset();
    props.history.goBack.mockReset();
  });

  it('should test the mapStateToProps method', () => {
    const state = {
      auth: {
        token: '123-456',
        enterpriseAccounts: [{} as EnterpriseAccount],
      } as AuthState,
      settings: {
        participating: false,
      } as SettingsState,
    } as AppState;

    const mappedProps = mapStateToProps(state);

    expect(mappedProps.hasMultipleAccounts).toBeTruthy();
    expect(mappedProps.settings.participating).toBeFalsy();
  });

  it('should render itself & its children', () => {
    const tree = TestRenderer.create(
      <MemoryRouter>
        <SettingsRoute {...props} />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });

  it('should press the logout', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Logout'));

    expect(props.logout).toHaveBeenCalledTimes(1);

    expect(ipcRenderer.send).toHaveBeenCalledTimes(1);
    expect(ipcRenderer.send).toHaveBeenCalledWith('update-icon');
    expect(props.history.goBack).toHaveBeenCalledTimes(1);
  });

  it('should call the componentWillReceiveProps method', function() {
    const { rerender } = render(<SettingsRoute {...props} />);

    expect(props.fetchNotifications).toHaveBeenCalledTimes(0);

    const updatedSettings = {
      ...props.settings,
      participating: !props.settings.participating,
    };

    rerender(<SettingsRoute {...props} settings={updatedSettings} />);

    expect(props.fetchNotifications).toHaveBeenCalledTimes(1);
  });

  it('should go back by pressing the icon', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);
    fireEvent.click(getByLabelText('Close Settings'));
    expect(props.history.goBack).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showOnlyParticipating checbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Show only participating'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the playSound checbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Play sound'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showNotifications checbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Show notifications'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the onClickMarkAsRead checbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('On Click, Mark as Read'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the openAtStartup checbox', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Open at startup'), {
      target: { checked: true },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
  });

  it('should toggle the showAppIcon radiogroup', () => {
    const { getByLabelText } = render(<SettingsRoute {...props} />);

    fireEvent.click(getByLabelText('Tray Icon'), {
      target: { value: 'tray' },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
    expect(props.updateSetting).toHaveBeenCalledWith('showAppIcon', 'tray');

    props.updateSetting.mockReset();

    fireEvent.click(getByLabelText('Dock Icon'), {
      target: { value: 'dock' },
    });

    expect(props.updateSetting).toHaveBeenCalledTimes(1);
    expect(props.updateSetting).toHaveBeenCalledWith('showAppIcon', 'dock');
  });
});
