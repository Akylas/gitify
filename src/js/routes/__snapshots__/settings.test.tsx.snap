// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/settings.tsx should render itself & its children 1`] = `
<div
  className="sc-EHOje bIfgFX"
>
  <div
    className="sc-bZQynM fISyMH"
  >
    <h3
      className="sc-dnqmqq jJLwtP"
    >
      Settings
    </h3>
    <button
      aria-label="Close Settings"
      className="sc-iwsKbI SeFqk"
      onClick={[Function]}
    >
      Ã—
    </button>
  </div>
  <div
    className="sc-gzVnrw kukodc"
  >
    <div
      className="sc-bdVaJa jfPzWn"
    >
      <label>
        <div
          className="sc-bwzfXH fnAEVU"
        >
          <input
            checked={false}
            className="sc-bxivhb hZSXJG"
            label="Show only participating"
            name="showOnlyParticipating"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="sc-ifAKCX kfOBSV"
          >
            <svg
              className="sc-htpNat iPrSGo"
              viewBox="0 0 24 24"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="1"
              >
                <g
                  stroke="#FFFFFF"
                  strokeWidth="3"
                  transform="translate(6.000000, 7.000000)"
                >
                  <polyline
                    points="13 0 4.0625 9 0 4.90909091"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
        <span
          style={
            Object {
              "marginLeft": "1rem",
            }
          }
        >
          Show only participating
        </span>
      </label>
    </div>
    <div
      className="sc-bdVaJa jfPzWn"
    >
      <label>
        <div
          className="sc-bwzfXH fnAEVU"
        >
          <input
            checked={true}
            className="sc-bxivhb hZSXJG"
            label="Play sound"
            name="playSound"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="sc-ifAKCX RDSID"
          >
            <svg
              className="sc-htpNat iPrSGo"
              viewBox="0 0 24 24"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="1"
              >
                <g
                  stroke="#FFFFFF"
                  strokeWidth="3"
                  transform="translate(6.000000, 7.000000)"
                >
                  <polyline
                    points="13 0 4.0625 9 0 4.90909091"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
        <span
          style={
            Object {
              "marginLeft": "1rem",
            }
          }
        >
          Play sound
        </span>
      </label>
    </div>
    <div
      className="sc-bdVaJa jfPzWn"
    >
      <label>
        <div
          className="sc-bwzfXH fnAEVU"
        >
          <input
            checked={true}
            className="sc-bxivhb hZSXJG"
            label="Show notifications"
            name="showNotifications"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="sc-ifAKCX RDSID"
          >
            <svg
              className="sc-htpNat iPrSGo"
              viewBox="0 0 24 24"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="1"
              >
                <g
                  stroke="#FFFFFF"
                  strokeWidth="3"
                  transform="translate(6.000000, 7.000000)"
                >
                  <polyline
                    points="13 0 4.0625 9 0 4.90909091"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
        <span
          style={
            Object {
              "marginLeft": "1rem",
            }
          }
        >
          Show notifications
        </span>
      </label>
    </div>
    <div
      className="sc-bdVaJa jfPzWn"
    >
      <label>
        <div
          className="sc-bwzfXH fnAEVU"
        >
          <input
            checked={false}
            className="sc-bxivhb hZSXJG"
            label="On Click, Mark as Read"
            name="onClickMarkAsRead"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="sc-ifAKCX kfOBSV"
          >
            <svg
              className="sc-htpNat iPrSGo"
              viewBox="0 0 24 24"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="1"
              >
                <g
                  stroke="#FFFFFF"
                  strokeWidth="3"
                  transform="translate(6.000000, 7.000000)"
                >
                  <polyline
                    points="13 0 4.0625 9 0 4.90909091"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
        <span
          style={
            Object {
              "marginLeft": "1rem",
            }
          }
        >
          On Click, Mark as Read
        </span>
      </label>
    </div>
    <div
      className="sc-bdVaJa jfPzWn"
    >
      <label>
        <div
          className="sc-bwzfXH fnAEVU"
        >
          <input
            checked={false}
            className="sc-bxivhb hZSXJG"
            label="Open at startup"
            name="openAtStartUp"
            onChange={[Function]}
            type="checkbox"
          />
          <div
            className="sc-ifAKCX kfOBSV"
          >
            <svg
              className="sc-htpNat iPrSGo"
              viewBox="0 0 24 24"
            >
              <g
                fill="none"
                fillRule="evenodd"
                stroke="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="1"
              >
                <g
                  stroke="#FFFFFF"
                  strokeWidth="3"
                  transform="translate(6.000000, 7.000000)"
                >
                  <polyline
                    points="13 0 4.0625 9 0 4.90909091"
                  />
                </g>
              </g>
            </svg>
          </div>
        </div>
        <span
          style={
            Object {
              "marginLeft": "1rem",
            }
          }
        >
          Open at startup
        </span>
      </label>
    </div>
  </div>
  <div
    className="sc-htoDjs bsVypt"
  >
    <small>
      Gitify v
      0.0.1
    </small>
    <div>
      <button
        aria-label="Login with GitHub Enterprise"
        className="sc-gZMcBi loTXCJ"
        onClick={[Function]}
      >
        Add Enterprise
      </button>
      <button
        aria-label="Logout"
        className="sc-gZMcBi loTXCJ"
        onClick={[Function]}
      >
        Logout
      </button>
      <button
        aria-label="Quit Gitify"
        className="sc-gZMcBi loTXCJ"
        onClick={[Function]}
      >
        Quit
      </button>
    </div>
  </div>
</div>
`;
